### DISCUSSION TITLE
Assert.IsType<T> not behaving quite as expected

### DISCUSSION THREAD

**Discussion by mwpowellhtx** [xunit/xunit#2665] [Question] [answered]
Maybe this has been addressed in current code base, I'm not sure.

At core I am evaluating `Assert.IsType<T>(...)` in my fluent assertions.

Trying to verify a web response is literally, `Assert.IsType<Task<Stream>>(...)`.

However, upon closer inspection of the actual types involved, they return something like `System.Runtime.CompilerServices.AsyncTaskMethodBuilder<...>`, which is something wholely other than that type.

But it does evaluate `true` using the following, `responseTask is Task<Stream>`, which is really peculiar when you pause to consider a `struct` value can be seen as a `class` instance. Unless there were some implicit type conversions going on there.

That aside, _xunit_ is trying to evaluate the type in the strictest sense, it seems, which fails, so this is failing, `responseType.AssertIsType<Task<Stream>>()`, seen [in my repository](https://github.com/mwpowellhtx/Xunit.Fluently/blob/master/src/xunit.fluently.assert/TypeAsserts.cs#L59), i.e. `InvokeUnaryWithReturn(Assert.IsType<T>, value)`, never mind my plumbing to make the fluent-isms work.

Alternatively, I can evaluate something such as the following, `responseTask.AssertTrue(x => x is Task<Stream>)`. Which is an acceptable workaround for now; but I think that `IsType<T>(...)` should probably allow for these sort of use cases.

At the moment I have the following dependencies involved, on `xunit.fluently.assert v1.7.0+` dependent upon `xunit.assert 2.4.1`, but perhaps this has already been addressed in a subsequent _xunit_ version, I don't know.

**Comment by mwpowellhtx** [comment, marked as answer]
This will work, is closer to what I am aiming 🎯 at I think. `responseTask.AssertIsAssignableFrom<Task<Stream>>()`

Which internally is, `InvokeUnaryWithReturn((Func<object, T>)Assert.IsAssignableFrom<T>, (object)value)`.

**Reply by bradwilson** [reply]
That is the correct way: using `Assert.IsAssignableFrom`. The documentation for both:

`Assert.IsType`: "Verifies that an object is exactly the given type (and not a derived type)."
`Assert.IsAssignableFrom`: "Verifies that an object is of the given type or a derived type."
